### PR TITLE
Expose saved queries in Project

### PR DESCRIPTION
**Pull request by pelavall**

### CODE REVIEW

File: src/Qwiq.Core.Rest/Project.cs [hunk 1]
```diff
@@ -64,7 +66,27 @@ internal Project([NotNull] TeamProjectReference project, [NotNull] WorkItemStore
...
+                new Lazy<IQueryFolderCollection>(() =>
+                {
+                    return new QueryFolderCollection(() =>
+                    {
+                        //BUGBUG: There is a bug in the GetQueryAsync in vsts where if a folder contains a '+' character it will return 404, even if the folder exists see here: https://developercommunity.visualstudio.com/content/problem/123660/when-a-saved-query-folder-contains-a-character-it.html
+                        QueryHierarchyItem FolderExpansionFunc(string path)
+                        {
+                            try
+                            {
+                                return store.NativeWorkItemStore.Value.GetQueryAsync(project.Id, path, QueryExpand.Wiql, 2).Result;
```
Comment (MattKotsenas) [inline src/Qwiq.Core.Rest/Project.cs]: Should `2` be turned into a constant to be self-documenting?

File: src/Qwiq.Core.Rest/QueryFolder.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Linq;
+using Microsoft.TeamFoundation.WorkItemTracking.WebApi.Models;
+
+namespace Microsoft.Qwiq.Client.Rest
+{
+    internal class QueryFolder : Qwiq.QueryFolder
+    {
+        internal QueryFolder(QueryHierarchyItem queryFolder, Func<string, QueryHierarchyItem> folderExpansionFunc)
```
Comment (MattKotsenas) [inline src/Qwiq.Core.Rest/QueryFolder.cs]: Looks like you're doing work in the constructor (calling `folderExpansionFunc`) that could `throw`.

File: src/Qwiq.Core.Rest/QueryFolder.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                new QueryFolderCollection(() =>
+                {
+                    if (queryFolder.Children == null)
+                    {
+                        return Enumerable.Empty<IQueryFolder>();
+                    }
+                    
+                    return queryFolder
+                        .Children
+                        .Where(q => q?.IsFolder != null && q.IsFolder.Value)
```
Comment (MattKotsenas) [inline src/Qwiq.Core.Rest/QueryFolder.cs]: Feels like this should be encapsulated?

Comment (MattKotsenas) [inline src/Qwiq.Core.Rest/QueryFolder.cs]: Couldn't this be shortened to `q?.IsFolder?.Value`?

File: src/Qwiq.Core/QueryDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public override int GetHashCode()
+        {
+            return QueryDefinitionComparer.Default.GetHashCode(this);
+        }
+
+        public override string ToString()
+        {
+            FormattableString s = $"{Id.ToString()} ({Name})";
```
Comment (MattKotsenas) [inline src/Qwiq.Core/QueryDefinition.cs]: nit: collapse

File: src/Qwiq.Core/QueryFolder.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public override int GetHashCode()
+        {
+            return QueryFolderComparer.Default.GetHashCode(this);
+        }
+
+        public override string ToString()
+        {
+            FormattableString s = $"{Id.ToString()} ({Name})";
```
Comment (MattKotsenas) [inline src/Qwiq.Core/QueryFolder.cs]: nit: collapse

File: test/Qwiq.Integration.Tests/Project/ProjectTests.cs [hunk 6]
```diff
@@ -47,6 +47,20 @@ public void Each_project_contains_the_same_Iteration_paths()
         [TestCategory("localOnly")]
         [TestCategory("SOAP")]
         [TestCategory("REST")]
+        public void Each_project_contains_the_same_saved_QueryHierarchy()
```
Comment (MattKotsenas) [inline test/Qwiq.Integration.Tests/Project/ProjectTests.cs:76]: Integration test but no unit tests

File: src/Qwiq.Core/IQueryFolderCollection.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+using System;
+
+namespace Microsoft.Qwiq
+{
+    public interface IQueryFolderCollection : IReadOnlyObjectWithIdCollection<IQueryFolder, Guid>, IEquatable<IQueryFolderCollection>
```
Comment (MattKotsenas) [inline src/Qwiq.Core/IQueryFolderCollection.cs:5]: Do we need to expose new types here over just `IReadOnlyCollection<T>`?

File: src/Qwiq.Core/Project.cs [hunk 8]
```diff
@@ -48,11 +47,18 @@ public bool Equals(IProject other)
 
         public IWorkItemTypeCollection WorkItemTypes => _wits.Value;
 
+        public IQueryFolderCollection QueryHierarchy => _queryHierarchy.Value;
```
Comment (MattKotsenas) [inline src/Qwiq.Core/Project.cs:55]: This name seems odd. Maybe `SavedQueries`?

File: src/Qwiq.Core/QueryDefinitionCollection.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Qwiq
+{
+    public class QueryDefinitionCollection : ReadOnlyObjectWithIdCollection<IQueryDefinition, Guid>, IQueryDefinitionCollection
```
Comment (MattKotsenas) [inline src/Qwiq.Core/QueryDefinitionCollection.cs:6]: Does this need to be its own type instead of just `ICollection<T>`?

File: src/Qwiq.Core.Soap/QueryFolder.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
+using System.Linq;
+using Microsoft.Qwiq.Exceptions;
+
+namespace Microsoft.Qwiq.Client.Soap
+{
+    public class QueryFolder : Qwiq.QueryFolder
+    {
+        public QueryFolder(TeamFoundation.WorkItemTracking.Client.QueryFolder queryFolder)
```
Comment (MattKotsenas) [inline src/Qwiq.Core.Soap/QueryFolder.cs]: Should this not be `public`? The constructor exposes the underlying TFS type.

File: src/Qwiq.Core.Soap/QueryDefinition.cs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+namespace Microsoft.Qwiq.Client.Soap
+{
+    public class QueryDefinition : Qwiq.QueryDefinition
+    {
+        public QueryDefinition(TeamFoundation.WorkItemTracking.Client.QueryDefinition queryDefinition)
```
Comment (MattKotsenas) [inline src/Qwiq.Core.Soap/QueryDefinition.cs]: Should this not be `public`? The constructor exposes the underlying TFS type.

File: src/Qwiq.Core.Rest/QueryDefinition.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+using Microsoft.TeamFoundation.WorkItemTracking.WebApi.Models;
+
+namespace Microsoft.Qwiq.Client.Rest
+{
+    public class QueryDefinition : Qwiq.QueryDefinition
+    {
+        public QueryDefinition(QueryHierarchyItem queryDefinition)
```
Comment (MattKotsenas) [inline src/Qwiq.Core.Rest/QueryDefinition.cs]: Should this not be `public`? The constructor exposes the underlying TFS type.

File: src/Qwiq.Core.Rest/Project.cs (hunk 1 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/Project.cs]: Yup, will do.

File: src/Qwiq.Core.Rest/QueryDefinition.cs (hunk 12 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryDefinition.cs]: Should be internal, will update.

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 2 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: It's not called in the constructor, it's called in a lazy func which is invoked later.

File: src/Qwiq.Core.Soap/QueryDefinition.cs (hunk 11 above)
Comment (pelavall) [reply src/Qwiq.Core.Soap/QueryDefinition.cs]: Should be internal, will update.

File: src/Qwiq.Core.Soap/QueryFolder.cs (hunk 10 above)
Comment (pelavall) [reply src/Qwiq.Core.Soap/QueryFolder.cs]: Should be internal, will update.

File: src/Qwiq.Core/IQueryFolderCollection.cs (hunk 7 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryFolderCollection.cs:5]: Yes to support look ups by Name and Id.

Comment (pelavall) [reply src/Qwiq.Core/IQueryFolderCollection.cs:5]: Indexing vs needing to use Single/First

File: src/Qwiq.Core/Project.cs (hunk 8 above)
Comment (pelavall) [reply src/Qwiq.Core/Project.cs:55]: QueryHierarchy is what vs calls it in the soap implementation, I thought about changing it to SavedQueries, but wasn't 100% interms of changing semi established naming. Still think we should change?

File: src/Qwiq.Core/QueryDefinition.cs (hunk 4 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryDefinition.cs]: will do.

File: src/Qwiq.Core/QueryDefinitionCollection.cs (hunk 9 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryDefinitionCollection.cs:6]: Yes to support indexing on name and id rather than needing to use single/first

File: src/Qwiq.Core/QueryFolder.cs (hunk 5 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryFolder.cs]: will do.

File: test/Qwiq.Integration.Tests/Project/ProjectTests.cs (hunk 6 above)
Comment (pelavall) [reply test/Qwiq.Integration.Tests/Project/ProjectTests.cs:76]: Will add.

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 3 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: Nope, nullable types don't support safe navigation operator, would also rely on null being falsy.

Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: Which part?

Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: updated

File: src/Qwiq.Core/QueryDefinition.cs (hunk 4 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryDefinition.cs]: updated

File: src/Qwiq.Core/QueryFolder.cs (hunk 5 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryFolder.cs]: updated

File: src/Qwiq.Core.Soap/QueryFolder.cs (hunk 10 above)
Comment (pelavall) [reply src/Qwiq.Core.Soap/QueryFolder.cs]: updated

File: src/Qwiq.Core.Soap/QueryDefinition.cs (hunk 11 above)
Comment (pelavall) [reply src/Qwiq.Core.Soap/QueryDefinition.cs]: updated

File: src/Qwiq.Core.Rest/QueryDefinition.cs (hunk 12 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryDefinition.cs]: updated

File: src/Qwiq.Core.Rest/Project.cs (hunk 1 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/Project.cs]: updated

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 3 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: Make this simpler by encapsulating logic checks into selfdocumenting extension methods

File: src/Qwiq.Core/Project.cs (hunk 8 above)
Comment (pelavall) [reply src/Qwiq.Core/Project.cs:55]: Change to SavedQueries

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 3 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: updated

File: src/Qwiq.Core.Rest/Project.cs [hunk 13]
```diff
@@ -64,7 +68,27 @@ internal Project([NotNull] TeamProjectReference project, [NotNull] WorkItemStore
...
+                    return new QueryFolderCollection(() =>
+                    {
+                        //BUGBUG: There is a bug in the GetQueryAsync in vsts where if a folder contains a '+' character it will return 404, even if the folder exists see here: https://developercommunity.visualstudio.com/content/problem/123660/when-a-saved-query-folder-contains-a-character-it.html
+                        QueryHierarchyItem FolderExpansionFunc(QueryHierarchyItem item)
+                        {
+                            try
+                            {
+                                return store.NativeWorkItemStore.Value.GetQueryAsync(project.Id, item.Path, QueryExpand.Wiql, MaxQueryFolderExpansionDepth).Result;
+                            }
+                            catch(VssServiceResponseException ex){
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/Project.cs]: Can't we inspect the `VssServiceResponseException` details and throw the appropriate message? The exception is thrown whenever any type of exception is encountered

File: src/Qwiq.Core.Rest/Extensions.cs [hunk 14]
```diff
@@ -40,5 +41,15 @@ internal static IInternalTeamProjectCollection AsProxy([CanBeNull] this VssConne
                        ? null
                        : ExceptionHandlingDynamicProxyFactory.Create<IInternalTeamProjectCollection>(new VssConnectionAdapter(tfsNative));
         }
+
+        internal static bool IsFolder([NotNull] this QueryHierarchyItem item)
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/Extensions.cs]: Missing annotatations

File: src/Qwiq.Core.Rest/QueryFolder.cs [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            : base(
+                queryFolder.Id,
+                queryFolder.Name,
+                new QueryFolderCollection(() =>
+                {
+                    return
+                        queryFolder
+                            .Children?
+                            .Where(q => q.IsFolder())
+                            .Select(q => !q.IsExpanded() ? folderExpansionFunc(q) : q)
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/QueryFolder.cs]: FYI: These lambdas allocate

File: src/Qwiq.Core.Rest/QueryFolder.cs [hunk 16]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Linq;
+using Microsoft.TeamFoundation.WorkItemTracking.WebApi.Models;
+
+namespace Microsoft.Qwiq.Client.Rest
+{
+    internal class QueryFolder : Qwiq.QueryFolder
+    {
+        internal QueryFolder(QueryHierarchyItem queryFolder, Func<QueryHierarchyItem, QueryHierarchyItem> folderExpansionFunc)
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/QueryFolder.cs]: `folderExpensionFunc` is nullable but not annotated

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 15 above)
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/QueryFolder.cs]: `folderExpansionFunc` can throw

File: src/Qwiq.Core.Rest/QueryDefinition.cs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+using Microsoft.TeamFoundation.WorkItemTracking.WebApi.Models;
+
+namespace Microsoft.Qwiq.Client.Rest
+{
+    internal class QueryDefinition : Qwiq.QueryDefinition
+    {
+        internal QueryDefinition(QueryHierarchyItem queryDefinition)
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/QueryDefinition.cs]: `queryDefinition` is nullable but not annotated

File: src/Qwiq.Core.Rest/QueryDefinition.cs [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+            : base(queryDefinition.Id, queryDefinition.Name, queryDefinition.Wiql)
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/QueryDefinition.cs]: Not checked for null, but used here resulting in a runtime `NullReferenceException`

File: src/Qwiq.Core/Project.cs [hunk 19]
```diff
@@ -48,11 +47,18 @@ public bool Equals(IProject other)
...
+
         public override bool Equals(object obj)
         {
             return ProjectComparer.Default.Equals(this, obj as IProject);
         }
 
+        public bool Equals(IProject other)
```
Comment (rjmurillo) [inline src/Qwiq.Core/Project.cs]: Is there a test for this?

File: src/Qwiq.Core/QueryDefinition.cs [hunk 20]
```diff
@@ -0,0 +1,39 @@
+using System;
+using System.Globalization;
+
+namespace Microsoft.Qwiq
+{
+    public class QueryDefinition : IQueryDefinition
+    {
+        public QueryDefinition(Guid id, string name, string queryText)
```
Comment (rjmurillo) [inline src/Qwiq.Core/QueryDefinition.cs]: `.ctor` in Core should not be public (we don't want folks creating types outside of the Mocks, REST, and SOAP clients)

Comment (rjmurillo) [inline src/Qwiq.Core/QueryDefinition.cs]: Add guards for parameters

File: src/Qwiq.Core/QueryFolder.cs [hunk 21]
```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Globalization;
+
+namespace Microsoft.Qwiq
+{
+    public class QueryFolder : IQueryFolder
+    {
+        public QueryFolder(Guid id, string name, IQueryFolderCollection subFolders, IQueryDefinitionCollection queries)
```
Comment (rjmurillo) [inline src/Qwiq.Core/QueryFolder.cs]: internal `.ctor`

Comment (rjmurillo) [inline src/Qwiq.Core/QueryFolder.cs]: Parameter guards

File: src/Qwiq.Core/IQueryFolder.cs [hunk 22]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace Microsoft.Qwiq
+{
+    public interface IQueryFolder : IIdentifiable<Guid>, IEquatable<IQueryFolder>
+    {
+        string Name { get; }
+        IQueryFolderCollection SubFolders { get; }
```
Comment (rjmurillo) [inline src/Qwiq.Core/IQueryFolder.cs:13]: This name doesn't make sense to me--the folder really is a folder hierarchy, and in a hierarchy, we have `IQueryFolderCollection` containing children, so I'd expect this to be more like the `INode` interface

File: src/Qwiq.Core/IQueryFolder.cs [hunk 23]
```diff
@@ -0,0 +1,11 @@
+using System;
+
+namespace Microsoft.Qwiq
+{
+    public interface IQueryFolder : IIdentifiable<Guid>, IEquatable<IQueryFolder>
```
Comment (rjmurillo) [inline src/Qwiq.Core/IQueryFolder.cs:6]: Don't we care about other metadata, like path and URI?

File: src/Qwiq.Core/IQueryFolder.cs [hunk 24]
```diff
@@ -0,0 +1,11 @@
...
+    {
+        string Name { get; }
+        IQueryFolderCollection SubFolders { get; }
+        IQueryDefinitionCollection SavedQueries { get; }
```
Comment (rjmurillo) [inline src/Qwiq.Core/IQueryFolder.cs:15]: Name doesn't make sense to me: there's no reason to quality the queries returned (e.g. they can't be anything else but saved). Expected to see something like `Queries` or `QueryCollection`

File: src/Qwiq.Core/IQueryDefinition.cs [hunk 25]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace Microsoft.Qwiq
+{
+    public interface IQueryDefinition : IIdentifiable<Guid>, IEquatable<IQueryDefinition>
+    {
+        string Name { get; }
+        string QueryText { get; }
```
Comment (rjmurillo) [inline src/Qwiq.Core/IQueryDefinition.cs]: WIQL?

File: src/Qwiq.Core/IQueryDefinition.cs [hunk 26]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace Microsoft.Qwiq
+{
+    public interface IQueryDefinition : IIdentifiable<Guid>, IEquatable<IQueryDefinition>
```
Comment (rjmurillo) [inline src/Qwiq.Core/IQueryDefinition.cs:7]: What about URI?

File: src/Qwiq.Core/IQueryFolder.cs (hunk 24 above)
Comment (rjmurillo) [reply src/Qwiq.Core/IQueryFolder.cs:15]: Saw that you and @MattKotsenas had a conversation about this in another file. The REST API structure is a hierarchy with limited vocabulary (e.g. a unidirected graph) and as such I expected the same types of names in the interface. That is, nodes have a name, are aware if they are a special type of node that can contain other nodes, and an enumeration of each special type of node (e.g. children). The SOAP object `QueryHierarchy` includes additional vocabulary, such as ancestor traversal, but relies on the enumerator to provide details of "children". Each child node is a descendant of `QueryItem` from what I can tell, making it less obvious than the REST API that there are children.

TL;DR - I expected names congruent with the data structure but did not find them. Great if there are "helpers" that perform certain types of common traversals automatically to keep things DRY for consumers. This new interface should be reconciled with `INode`

File: src/Qwiq.Core/QueryDefinition.cs (hunk 20 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryDefinition.cs]: Should it be marked as abstract then? Either way will mark as internal and ensure a mock implementation

File: src/Qwiq.Core/IQueryFolder.cs (hunk 23 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryFolder.cs:6]: I don't think so? At least, we don't have a usecase for it yet, and I couldn't easily think of a good one, but that just might be lack of imagination on my part 😃

File: src/Qwiq.Core/QueryDefinition.cs (hunk 20 above)
Comment (pelavall) [reply src/Qwiq.Core/QueryDefinition.cs]: will do.

File: src/Qwiq.Core/IQueryFolder.cs (hunk 22 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryFolder.cs:13]: I'll take a look and update appropriately

File: src/Qwiq.Core/IQueryFolder.cs (hunk 24 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryFolder.cs:15]: will update

File: src/Qwiq.Core/IQueryDefinition.cs (hunk 26 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryDefinition.cs:7]: I'll add uri.

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 16 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: will update

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 15 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: Would establishing private methods to perform the checks be better?

File: src/Qwiq.Core/Project.cs (hunk 19 above)
Comment (pelavall) [reply src/Qwiq.Core/Project.cs]: There are integration tests which use equals, but no unit tests

File: src/Qwiq.Core/IQueryFolder.cs (hunk 23 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryFolder.cs:6]: I'll add path as that is common to both soap and rest, but Uri doesn't exist in Soap. We could support it by adding a new IQueryFolder to Rest which includes the added param? Or add it to soap with a NotSupportedException

File: src/Qwiq.Core/IQueryDefinition.cs (hunk 25 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryDefinition.cs]: Sure, I was following the soap naming convention, which is QueryText, but Wiql is probably more clear.

File: src/Qwiq.Core/IQueryDefinition.cs (hunk 26 above)
Comment (pelavall) [reply src/Qwiq.Core/IQueryDefinition.cs:7]: Err...So Uri doesn't exist for soap, we can either do NotSupported for SOAP or add additional IQueryDefinition for Rest with Rest only property.

File: src/Qwiq.Core.Rest/Project.cs (hunk 13 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/Project.cs]: For the bug, yes, I'll update the logic to peform the checks and throw a more specific error. Otherwise I don't know we can provide any additional details than the inner exception provides.

Comment (rjmurillo) [reply src/Qwiq.Core.Rest/Project.cs]: @pelavall You're right. If you would inspect the error to throw the exception you have (which is super helpful), ensuring what we throw is both helpful and relevant. In other cases, throw the caught exception since we can't add any value

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 15 above)
Comment (rjmurillo) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: @pelavall Local/Named/Lambda doesn't matter here--the LINQ portion is performing the allocation. This also appears to be done recursively here when the collection is first enumerated. Do you know what the perf characteristics of this are? More specifically, SOAP vs REST

File: src/Qwiq.Core.Rest/Project.cs [hunk 27]
```diff
@@ -64,7 +68,27 @@ internal Project([NotNull] TeamProjectReference project, [NotNull] WorkItemStore
...
+                new Lazy<IQueryFolderCollection>(() =>
+                {
+                    return new QueryFolderCollection(() =>
+                    {
+                        //BUGBUG: There is a bug in the GetQueryAsync in vsts where if a folder contains a '+' character it will return 404, even if the folder exists see here: https://developercommunity.visualstudio.com/content/problem/123660/when-a-saved-query-folder-contains-a-character-it.html
+                        QueryHierarchyItem FolderExpansionFunc(QueryHierarchyItem item)
+                        {
+                            try
+                            {
+                                return store.NativeWorkItemStore.Value.GetQueryAsync(project.Id, item.Path, QueryExpand.Wiql, MaxQueryFolderExpansionDepth).Result;
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/Project.cs]: Please review method for awaiting for and obtaining result (s.g. L67 L68)

File: src/Qwiq.Core.Rest/Project.cs [hunk 28]
```diff
@@ -64,7 +68,27 @@ internal Project([NotNull] TeamProjectReference project, [NotNull] WorkItemStore
...
 
                             return new Node(result).ChildNodes;
-                        }))
+                        }),
+                new Lazy<IQueryFolderCollection>(() =>
+                {
+                    return new QueryFolderCollection(() =>
+                    {
+                        //BUGBUG: There is a bug in the GetQueryAsync in vsts where if a folder contains a '+' character it will return 404, even if the folder exists see here: https://developercommunity.visualstudio.com/content/problem/123660/when-a-saved-query-folder-contains-a-character-it.html
+                        QueryHierarchyItem FolderExpansionFunc(QueryHierarchyItem item)
```
Comment (rjmurillo) [inline src/Qwiq.Core.Rest/Project.cs]: Why does this need to be passed into the ctor? Can we encapsulate it in the Rests's `QueryFolder` object? Unclear why `Project` would need to know this information

File: src/Qwiq.Core/IQueryFolder.cs (hunk 23 above)
Comment (rjmurillo) [reply src/Qwiq.Core/IQueryFolder.cs:6]: @pelavall I'm fine with omitting it since it has no universal application to both (a URI can be constructed for SOAP, but QWIQ must construct it, rather than having it supplied). You were right: YAGNI will apply here

File: src/Qwiq.Core.Rest/QueryFolder.cs (hunk 15 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: REST is significantly slower. With enumerating the full set of WPT queries it's about 17 seconds(REST) vs 7 seconds(SOAP). The construction of the "first layer" is 1 second(REST) vs 7 seconds(SOAP), for the getting the full depth, through gethashcode, it's an additional, 16 seconds(REST) vs 34 ms(SOAP). Looks like on access of QueryHierarchy in SOAP the full structure is retrieved, where as with rest each request is only 2 deep. Each call to the rest query api is between 200ms and 600ms. I've updated the integration test to include the perf.

Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: So we can't really improve the speed for iterating, but if it's a concern, I can reimplement QueryFolderCollection/QueryDefinitionCollection to not be backed by ReadOnlyObjectCollection or update the implementation of ReadOnlyObjectCollection to support getting individual element without iterating.

Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: For getting a specific query it looks like it's about 8 seconds in SOAP and 12 seconds in REST

Comment (pelavall) [reply src/Qwiq.Core.Rest/QueryFolder.cs]: So, actually, if we update the implementation we could potentially make REST faster for retrieving specific queries, but for retrieving the full tree, it seems like SOAP will be faster.

File: src/Qwiq.Core.Rest/Project.cs (hunk 28 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/Project.cs]: I'll revisit

File: src/Qwiq.Core.Rest/Project.cs (hunk 27 above)
Comment (pelavall) [reply src/Qwiq.Core.Rest/Project.cs]: Changed to GetAwaiter().GetResult

**Comment by rjmurillo** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/lecantaloop/qwiq) found an issue: The constructor is present to ensure proper initialization by CLR. This can be ignored.

It's currently on:
[src\Qwiq.Core\QueryFolderComparer.cs:55
](https://github.com/LeCantaloop/Qwiq/blob/bd33ac587e5dc490be546c3f75d6e6637d447264/src/Qwiq.Core/QueryFolderComparer.cs#L55)Commit bd33ac587e5dc490be546c3f75d6e6637d447264

**Comment by rjmurillo** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/lecantaloop/qwiq) found an issue: The constructor is present to ensure proper initialization by CLR. This can be ignored.

It's currently on:
[src\Qwiq.Core\QueryDefinitionComparer.cs:51
](https://github.com/LeCantaloop/Qwiq/blob/bd33ac587e5dc490be546c3f75d6e6637d447264/src/Qwiq.Core/QueryDefinitionComparer.cs#L51)Commit bd33ac587e5dc490be546c3f75d6e6637d447264
